tools/capable.py

```diff
@@ -86,25 +92,48 @@
 #include <linux/sched.h>
 
 struct data_t {
-   // switch to u32s when supported
-   u64 pid;
-   u64 uid;
+   u64 pid_tgid;
+   u32 tgid;
+   u32 pid;
+   u32 uid;
    int cap;
    int audit;
    char comm[TASK_COMM_LEN];
+#ifdef KERNEL_STACKS
+   int kernel_stack_id;
+#endif
+#ifdef USER_STACKS
+   int user_stack_id;
+#endif
+#if ((defined(KERNEL_STACKS) && defined(USER_STACKS)) || (!defined(KERNEL_STACKS) && !defined(USER_STACKS)))
+   int __pad;
+#endif
```

Comment:
I think line 108-110 here is not needed. Both bpf program and python side should handle padding properely.

Reply:
It's not needed anymore if I drop pid_tgid from the struct.
However, if I keep it, I guess there are some padding / alignment issues, because I get this error:
```
$ sudo ./tools/capable.py 
bpf: Failed to load program: Permission denied
0: (bf) r6 = r1
1: (79) r8 = *(u64 *)(r6 +96)
2: (79) r9 = *(u64 *)(r6 +88)
3: (85) call bpf_get_current_pid_tgid#14
4: (bf) r7 = r0
5: (bf) r1 = r9
6: (67) r1 <<= 32
7: (77) r1 >>= 32
8: (15) if r1 == 0x0 goto pc+31
 R0=inv(id=0) R1=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv(id=0) R9=inv(id=0) R10=fp0,call_-1
9: (bf) r1 = r7
10: (67) r1 <<= 32
11: (77) r1 >>= 32
12: (15) if r1 == 0x5cba goto pc+27
 R0=inv(id=0) R1=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=inv(id=0) R8=inv(id=0) R9=inv(id=0) R10=fp0,call_-1
13: (85) call bpf_get_current_uid_gid#15
14: (b7) r1 = 0
15: (7b) *(u64 *)(r10 -48) = r1
16: (bf) r2 = r7
17: (77) r2 >>= 32
18: (63) *(u32 *)(r10 -40) = r2
19: (63) *(u32 *)(r10 -36) = r7
20: (63) *(u32 *)(r10 -32) = r0
21: (63) *(u32 *)(r10 -28) = r8
22: (63) *(u32 *)(r10 -24) = r9
23: (63) *(u32 *)(r10 -8) = r1
24: (63) *(u32 *)(r10 -12) = r1
25: (63) *(u32 *)(r10 -16) = r1
26: (63) *(u32 *)(r10 -20) = r1
27: (bf) r1 = r10
28: (07) r1 += -20
29: (b7) r2 = 16
30: (85) call bpf_get_current_comm#16
31: (18) r2 = 0xffff8802227c4200
33: (bf) r4 = r10
34: (07) r4 += -48
35: (bf) r1 = r6
36: (18) r3 = 0xffffffff
38: (b7) r5 = 48
39: (85) call bpf_perf_event_output#25
invalid indirect read from stack off -48+44 size 48

Traceback (most recent call last):
  File "./tools/capable.py", line 157, in <module>
    b = BPF(text=bpf_text)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 337, in __init__
    self._trace_autoload()
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 1030, in _trace_autoload
    fn = self.load_func(func_name, BPF.KPROBE)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 377, in load_func
    (func_name, errstr))
Exception: Failed to load BPF program kprobe__cap_capable: Permission denied
```
But, dropping pid_tgid from the struct also solves this problem, so I'll drop also __pad.